File: Luogu/P_2455_SDOI_2006_线性方程组.py
n = int(input())
mat = [list(map(float, input().split())) for _ in range(n)]

def gauss_elimination(mat, ans, eps=1e-6):
    # Forward elimination
    rank = 0
    for col in range(n):
        # Find pivot
        pivot = rank
        for i in range(rank + 1, n):
            if abs(mat[i][col]) > abs(mat[pivot][col]):
                pivot = i
        
        # If current column is all zeros, continue to next column
        if abs(mat[pivot][col]) < eps:
            continue
        
        # Swap rows if necessary
        if pivot != rank:
            mat[rank], mat[pivot] = mat[pivot], mat[rank]
        
        # Eliminate column
        for i in range(rank + 1, n):
            if abs(mat[i][col]) < eps:
                continue
            factor = mat[i][col] / mat[rank][col]
            for k in range(col, n + 1):
                mat[i][k] -= mat[rank][k] * factor
        
        rank += 1
    
    for i in range(rank, n):
        if abs(mat[i][n]) > eps:
            return -1  # No solution
    
    if rank < n:
        return 0  # Infinite solutions
    
    # Back substitution
    for i in range(n - 1, -1, -1):
        sum_val = sum(mat[i][j] * ans[j] for j in range(i + 1, n))
        ans[i] = (mat[i][n] - sum_val) / mat[i][i]
        if abs(ans[i]) < eps:
            ans[i] = 0
    return 1  # Unique solution

ans = [0] * n
flag = gauss_elimination(mat, ans)

if flag == -1:
    print(-1)  # No solution
elif flag == 0:
    print(0)   # Infinite solutions
else:
    for i, x in enumerate(ans):
        print(f"x{i+1}={x:.2f}")

File: Luogu/P_2455_SDOI_2006_线性方程组.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

// Gauss elimination solver
int gauss_elimination(vector<vector<double>>& mat, vector<double>& ans, const double eps = 1e-6) {
    int n = mat.size(), rank = 0;
    
    // Forward elimination
    for (int col = 0; col < n; col++) {
        // Find pivot
        int pivot = rank;
        for (int i = rank + 1; i < n; i++) {
            if (abs(mat[i][col]) > abs(mat[pivot][col])) {
                pivot = i;
            }
        }
        
        // If current column is all zeros, continue
        if (abs(mat[pivot][col]) < eps) continue;
        
        // Swap rows if necessary
        if (pivot != rank) {
            swap(mat[rank], mat[pivot]);
        }
        
        // Eliminate column
        for (int i = rank + 1; i < n; i++) {
            if (abs(mat[i][col]) < eps) continue;
            double factor = mat[i][col] / mat[rank][col];
            for (int k = col; k <= n; k++) {
                mat[i][k] -= mat[rank][k] * factor;
            }
        }
        rank++;
    }
    
    // Check for no solution
    for (int i = rank; i < n; i++) {
        if (abs(mat[i][n]) > eps) return -1;  // No solution
    }
    
    // Check for infinite solutions
    if (rank < n) return 0;  // Infinite solutions
    
    // Back substitution
    for (int i = n - 1; i >= 0; i--) {
        double sum = 0;
        for (int j = i + 1; j < n; j++) {
            sum += mat[i][j] * ans[j];
        }
        ans[i] = (mat[i][n] - sum) / mat[i][i];
        if (abs(ans[i]) < eps) ans[i] = 0;
    }
    
    return 1;  // Unique solution
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n; cin >> n;
    
    vector<vector<double>> mat(n, vector<double>(n + 1));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j <= n; j++) {
            cin >> mat[i][j];
        }
    }
    
    vector<double> ans(n, 0);
    int flag = gauss_elimination(mat, ans);
    
    if (flag == -1) {
        cout << -1 << endl;  // No solution
    } else if (flag == 0) {
        cout << 0 << endl;   // Infinite solutions
    } else {
        for (int i = 0; i < n; i++) {
            cout << format("x{}={:.2f}", i + 1, ans[i]) << endl;
        }
    }
    
    return 0;
}